streams/src/main/java/org/apache/kafka/streams/kstream/internals/BranchedKStreamImpl.java

```diff
@@ -0,0 +1,97 @@
+
+    private static final String BRANCH_NAME = "KSTREAM-BRANCH-";
+
+    private final KStreamImpl<K, V> source;
+    private final boolean repartitionRequired;
+    private final String splitterName;
+    private final Map<String, KStream<K, V>> result = new HashMap<>();
+
+    private final List<Predicate<? super K, ? super V>> predicates = new ArrayList<>();
+    private final List<String> childNames = new ArrayList<>();
+    private final ProcessorGraphNode<K, V> splitterNode;
+
+    BranchedKStreamImpl(final KStreamImpl<K, V> source, final boolean repartitionRequired, final NamedInternal named) {
+        this.source = source;
+        this.repartitionRequired = repartitionRequired;
+        this.splitterName = named.orElseGenerateWithPrefix(source.builder, BRANCH_NAME);
+
+        // predicates and childNames are passed by reference so when the user adds a branch they get added to
+        final ProcessorParameters<K, V> processorParameters =
+                new ProcessorParameters<>(new KStreamBranch<>(predicates, childNames), splitterName);
+        splitterNode = new ProcessorGraphNode<>(splitterName, processorParameters);
+        source.builder.addGraphNode(source.streamsGraphNode, splitterNode);
+    }
+
+    @Override
+    public BranchedKStream<K, V> branch(final Predicate<? super K, ? super V> predicate) {
+        return branch(predicate, BranchedInternal.empty());
+    }
+
+    @Override
+    public BranchedKStream<K, V> branch(final Predicate<? super K, ? super V> predicate, final Branched<K, V> branched) {
+        predicates.add(predicate);
+        createBranch(branched, predicates.size());
+        return this;
+    }
+
+    @Override
+    public Map<String, KStream<K, V>> defaultBranch() {
+        return defaultBranch(BranchedInternal.empty());
+    }
+
+    @Override
+    public Map<String, KStream<K, V>> defaultBranch(final Branched<K, V> branched) {
+        createBranch(branched, 0);
+        return result;
+    }
+
+    private void createBranch(final Branched<K, V> branched, final int index) {
+        final BranchedInternal<K, V> branchedInternal = new BranchedInternal<>(branched);
+        final String branchChildName = branchedInternal.branchProcessorName(splitterName, index);
+        childNames.add(branchChildName);
+        source.builder.newProcessorName(branchChildName);
+        final ProcessorParameters<K, V> parameters = new ProcessorParameters<>(new PassThrough<>(), branchChildName);
+        final ProcessorGraphNode<K, V> branchChildNode = new ProcessorGraphNode<>(branchChildName, parameters);
+        source.builder.addGraphNode(splitterNode, branchChildNode);
+        final KStreamImpl<K, V> newStream = new KStreamImpl<>(branchChildName, source.keySerde,
+                source.valueSerde, source.subTopologySourceNodes,
+                repartitionRequired, branchChildNode, source.builder);
+        branchedInternal.process(newStream, branchChildName, result);

```

Comment:
Same comment as for `branchProcessorName` -- it seems better to keep `BranchedInternal` a simple container and pull in the code into this method.

Reply:
done